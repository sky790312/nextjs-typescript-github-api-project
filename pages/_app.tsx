import { GlobalStyles, theme } from '@/GlobalStyles'
import { ThemeProvider } from 'styled-components'

export default function App({ Component }) {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      <Component />
    </ThemeProvider>
  )
}
