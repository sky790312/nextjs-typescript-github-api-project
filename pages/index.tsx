import React, { useState, useEffect, useRef, useCallback } from 'react'
import styled from 'styled-components'
import Head from 'next/head'
import { Container, TextCenterContainer } from '@/GlobalStyles'
import { SearchSection } from '@/components/SearchSection'
import { List } from '@/components/List'
import { Spinner } from '@/components/uiComponents/Spinner'
import { fetchGithubRepo } from '@/api'
import {
  FetchGithubRepoParams,
  FetchGithubRepoRespSchema,
  GithubRepoSchema,
} from '@/api/schema'
import {
  DEFAULT_PER_PAGES,
  AUTO_LOAD_MORE_DELAY_TIME,
  DEBOUNCE_DELAY_TIME,
} from '@/constants'
import { isClient, debounce } from '@/utils'

export const Home = React.memo(
  (): JSX.Element => {
    const [repos, setRepos] = useState<GithubRepoSchema[]>([])
    const [totalPages, setTotalPages] = useState(0)
    const page = useRef(0)
    const keyword = useRef('')

    const [element, setElement] = useState(null)
    const [isLoading, setIsLoading] = useState(false)
    const prevY = useRef(0)

    const perPages = DEFAULT_PER_PAGES
    const debounceDelayTime = DEBOUNCE_DELAY_TIME

    const observer = useRef(
      isClient() &&
        new IntersectionObserver(
          entries => {
            const firstEntry = entries[0]
            const y = firstEntry.boundingClientRect.y

            if (prevY.current > y) {
              setTimeout(() => loadMore(), AUTO_LOAD_MORE_DELAY_TIME)
            }

            prevY.current = y
          },
          { threshold: 0 },
        ),
    )

    useEffect(() => {
      const currentElement = element
      const currentObserver = observer.current

      if (currentElement) {
        currentObserver.observe(currentElement)
      }

      return () => {
        if (currentElement) {
          currentObserver.unobserve(currentElement)
        }
      }
    }, [element])

    const fetchData = useCallback(async () => {
      setIsLoading(true)
      try {
        const payload: FetchGithubRepoParams = {
          q: keyword.current,
          page: page.current,
          // eslint-disable-next-line @typescript-eslint/camelcase
          per_page: perPages,
        }
        const data: FetchGithubRepoRespSchema = await fetchGithubRepo(payload)
        setIsLoading(false)
        return { data }
      } catch (error) {
        setIsLoading(false)
        return { hasError: true }
      }
    }, [perPages])

    const loadMore = async () => {
      if (isLoading) {
        return
      }

      if (page.current < totalPages) {
        return
      }

      page.current++
      const { data, hasError } = await fetchData()
      if (hasError) {
        return
      }

      setRepos(prevRepos => [...prevRepos, ...data.items])
    }

    const handleKeyword = useCallback(
      debounce(async (newKeyword: string) => {
        keyword.current = newKeyword
        page.current = 1

        const { data, hasError } = await fetchData()
        if (hasError) {
          return
        }

        setRepos(data.items)
        const pages = Math.ceil(data.total_count / DEFAULT_PER_PAGES)
        setTotalPages(pages)
      }, debounceDelayTime),
      [fetchData],
    )

    return (
      <>
        <Head>
          <title>Create Next App</title>
          <link rel='icon' href='/favicon.ico' />
        </Head>
        <Container>
          <SearchSection handleKeyword={handleKeyword} />
          <List items={repos} />
          <LoadingContainer
            ref={setElement}
            shouldHide={page.current >= totalPages}
          >
            {isLoading && <Spinner />}
            <Button disabled={isLoading}>Load More</Button>
          </LoadingContainer>
        </Container>
      </>
    )
  },
)

export default Home

const LoadingContainer = styled(TextCenterContainer)<{ shouldHide: boolean }>`
  position: relative;
  padding: 10px;
  visibility: ${props => (props.shouldHide ? 'hidden' : 'visible')};
`

const Button = styled.button<{ disabled: boolean }>`
  opacity: ${props => (props.disabled ? '0.3' : '1')};
  cursor: ${props => (props.disabled ? 'default' : 'pointer')};
`
