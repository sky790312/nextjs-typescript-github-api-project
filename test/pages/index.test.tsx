import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import { Home } from '@/pages/index'
import { setupIntersectionObserverMock } from '@/test/__mocks__/intersectionObserverMock'

beforeEach(() => {
  setupIntersectionObserverMock()
})

describe('Home page', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(withProviders(<Home />), {})
    expect(asFragment()).toMatchSnapshot()
  })
})
