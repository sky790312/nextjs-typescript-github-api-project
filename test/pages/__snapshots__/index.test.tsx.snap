// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="GlobalStyles__Container-sc-1m5drgn-0 frTxMm"
  >
    <div
      class="GlobalStyles__TextCenterContainer-sc-1m5drgn-1 kwExyG"
    >
      <h3>
        search github repo by keyowrd: 
      </h3>
      <input
        type="text"
      />
    </div>
    <div
      class="List__ListContainer-sc-1l9sj4l-0 entigL"
    />
    <div
      class="GlobalStyles__TextCenterContainer-sc-1m5drgn-1 pages__LoadingContainer-sc-1cnstz1-0 kwExyG fQicZO"
    >
      <button
        class="pages__Button-sc-1cnstz1-1 iMUhti"
      >
        Load More
      </button>
    </div>
  </div>
</DocumentFragment>
`;
